File: tensor_processing.R
library(dplyr)

# Read in counts, leaving sample name column as part of the df
counts <- read.csv(snakemake@input[['counts']], sep='\t')
sample_info <- read.csv(snakemake@input[['sample_info']], sep='\t')

# Sort the samples by cell type and then individual
# (The samples are already sorted like this I think, but best to be safe)
# Discard NK cells as this cell type isn't complete
sorted_samples <- sample_info %>%
    filter(FactorValue..cell.type. != 'NK') %>%
    arrange(FactorValue..cell.type., Characteristics..individual.)

# Pick out the samples in the required order
sorted_counts <- sorted_samples %>%
    select(Comment..ENA_SAMPLE.) %>%
    left_join(sorted_sample_names,
              counts,
              by=c('Comment..ENA_SAMPLE.' = 'X'))

# Write out sorted sample info
write.table(sorted_samples, file = snakemake@output[["sample_info"]], row.names = FALSE, col.names = TRUE, sep="\t")

# Write out sorted counts matrix, without gene names or sample names
write.table(sorted_counts[, -1], file = snakemake@output[["Y"]], row.names = FALSE, col.names = FALSE, sep="\t")
writeLines(colnames(sorted_counts[, -1]), con = snakemake@output[["gene_names"]])

# Write out number of individuals (N), needed by tensor methods
N <- length(unique(sorted_samples$Characteristics..individual.))
writeLines(as.character(), con = snakemake@output[["N"]])

File: tensor_processing.py
import pandas as pd
import logging

def convert_to_tensor_dataset(Y_file,
                              tensor_Y_file,
                              sample_info_file,
                              tensor_sample_info_file,
                              N_file):
    sample_info = pd.read_csv(sample_info_file, sep='\t')
    Y = pd.read_csv(Y_file,
                    sep='\t',
                    header=None)
    logging.info(sample_info)
    logging.info(Y.shape)
    logging.info(sample_info.shape)

    # Sort by cell type, then individual
    sorted_sample_info = sample_info.sort_values(['FactorValue..cell.type.',
                                                  'Characteristics..individual.'])
    tensor_sample_info = sorted_sample_info[sorted_sample_info['FactorValue..cell.type.'] != 'NK']
    tensor_samples = tensor_sample_info['Comment..ENA_SAMPLE.']
    logging.info(tensor_sample_info)
    logging.info(f"Should have 120 tensor samples, have {len(tensor_samples)}")

    sample_info.index = sample_info['Comment..ENA_SAMPLE.']
    row_indices = [sample_info.index.get_loc(sample_id)
                   for sample_id in tensor_samples]


    tensor_Y = Y.iloc[row_indices, :]
    tensor_Y.to_csv(tensor_Y_file, sep='\t', header=False, index=False)

    tensor_sample_info.to_csv(tensor_sample_info_file, sep='\t', index=False)

    N = tensor_sample_info['FactorValue..cell.type.'].nunique()
    logging.info(f"Should be 6 cell types, found {N}")

    with open(N_file, 'w') as f:
        f.write(str(N))

